components/butterfly/src/rumor/service_health.rs

```diff
@@ -0,0 +1,131 @@
+                        service_group,
+                        incarnation: 0,
+                        health }
+    }
+}
+
+impl fmt::Display for ServiceHealth {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f,
+               "ServiceHealth i/{} m/{} sg/{} h/{}",
+               self.incarnation, self.member_id, self.service_group, self.health)
+    }
+}
+
+impl PartialOrd for ServiceHealth {
+    fn partial_cmp(&self, other: &ServiceHealth) -> Option<Ordering> {
+        if self.member_id != other.member_id && self.service_group != other.service_group {
+            None
+        } else {
+            Some(self.incarnation.cmp(&other.incarnation))
+        }
+    }
+}
+
+impl FromProto<ProtoRumor> for ServiceHealth {
+    fn from_proto(rumor: ProtoRumor) -> Result<Self> {
+        let payload = match rumor.payload.ok_or(Error::ProtocolMismatch("payload"))? {
+            RumorPayload::ServiceHealth(payload) => payload,
+            _ => panic!("from-bytes service-health"),
+        };
+        Ok(ServiceHealth { member_id:     rumor.from_id
+                                               .ok_or(Error::ProtocolMismatch("from-id"))?,
+                           service_group:
+                               payload.service_group
+                                      .ok_or(Error::ProtocolMismatch("service-group"))
+                                      .and_then(|s| {
+                                          ServiceGroup::from_str(&s).map_err(Error::from)
+                                      })?,
+                           incarnation:   payload.incarnation.unwrap_or(0),
+                           health:        payload.health
+                                                 .ok_or(Error::ProtocolMismatch("health"))?
+                                                 .into(), })
+    }
+}
+
+impl From<ServiceHealth> for newscast::ServiceHealth {
+    fn from(value: ServiceHealth) -> Self {
+        newscast::ServiceHealth { member_id:     Some(value.member_id),
+                                  service_group: Some(value.service_group.to_string()),
+                                  incarnation:   Some(value.incarnation),
+                                  health:        Some(value.health.into()), }
+    }
+}
+
+impl protocol::Message<ProtoRumor> for ServiceHealth {
+    const MESSAGE_ID: &'static str = "ServiceHealth";
+}
+
+impl Rumor for ServiceHealth {
+    /// Follows a simple pattern; if we have a newer incarnation than the one we already have, the
+    /// new one wins. So far, these never change.
+    fn merge(&mut self, mut other: ServiceHealth) -> bool {
+        if *self >= other {
+            false
+        } else {
+            mem::swap(self, &mut other);
+            true
+        }
+    }
+
+    fn kind(&self) -> RumorType { RumorType::ServiceHealth }
+
+    fn id(&self) -> &str { &self.member_id }
+
+    fn key(&self) -> &str { &self.service_group }
```

Comment:
:thinking: Since health of a service is, thus far, a Supervisor-specific thing (Service X on Supervisor A can be unhealthy, while Service X on Supervisor B can be just fine), this starts to feel like we'll need to have logic to coalesce the health of each service instance into an overall "service health", which could get really thorny :grimacing: